src/pybind/mgr/mgr_module.py

```diff
@@ -1070,7 +1070,14 @@ def get_metadata(self, svc_type, svc_id):
             calling this
         :rtype: dict, or None if no metadata found
         """
-        return self._ceph_get_metadata(svc_type, svc_id)
+        metadata = self._ceph_get_metadata(svc_type, svc_id)
+        if metadata is None:
+            if default is None:
+                return None
+            else:
+                return defaultdict(lambda: default)
```

Comment:
```suggestion
        if metadata is None:
            return defaultdict(lambda: default)
```

Reply:
still not completely right .. 

I want to make the suggested change, but then we need to modify all ref's that look like this:
```py
if get_metadata is None:
   ...
````

otherwise, the status module would need to pass an empty str (or similar):
```py
metadata = self.get_metadata('mds', info['name'], default='')
mds_versions[metadata.get('ceph_version', 'unknown')].append(info['name'])
```

@tchaikov  wdyt?

Reply:
@mgfritch i think your original fix is to provide a default value of "unknown" for "ceph_version".

as we have
```py
metadata = self.get_metadata('mds', info['name'])
mds_versions[metadata.get('ceph_version', "unknown")].append(info['name'])
```

when trying to fill the mdsmap with mds daemon's version.

with the suggested change, i think we can instead put something like
```py
metadata = self.get_metadata('mds', info['name'], 'unknown')
mds_versions[metadata.get('ceph_version')].append(info['name'])
```
which works even if the metadata of `f'mds.info['name']` is not available yet.

because we are only interested in a single value in metadata, a singe `default` value would suffice. if we want to read more than one value, and would like to have different default values, then `default` should be a dictionary, or we need to use `kwargs` for passing the default values.

> I want to make the suggested change, but then we need to modify all ref's that look like this:
> ```py
> if get_metadata is None:
>    ...
> ````

i don't think it's necessary, as, if `default` is `None`, the behavior of the new `self.get_metadata()` is identical to that of the existing one.

Reply:
> with the suggested change, i think we can instead put something like
> 
> ```python
> metadata = self.get_metadata('mds', info['name'], 'unknown')
> mds_versions[metadata.get('ceph_version')].append(info['name'])
> ```
> 
> which works even if the metadata of `f'mds.info['name']` is not available yet.
> 

almost works but there are interesting nuances with the default value of an unknown key:
```py
>>> from collections import defaultdict
>>> d = defaultdict(lambda: 'unknown')
>>> d.get('foo') is None
True
>>> d['foo']
'unknown'
```

> because we are only interested in a single value in metadata, a singe `default` value would suffice. if we want to read more than one value, and would like to have different default values, then `default` should be a dictionary, or we need to use `kwargs` for passing the default values.
> 

I moved the defaultdict out of the `get_metadata` func (https://github.com/ceph/ceph/pull/35161/commits/f06e60b591d7921b0afdd40219f3866773b00683)

not really sure if this is any cleaner, but any additional thoughts/feedback is welcome